packages/klighd-core/src/diagram-server.ts

```diff
@@ -0,0 +1,232 @@
+
+        const wasDiagramModelUpdated =
+            message.action.kind === SetModelCommand.KIND ||
+            message.action.kind === KlighdUpdateModelAction.KIND;
+
+        if (wasDiagramModelUpdated && this.preferencesRegistry.preferences.resizeToFit) {
+            this.actionDispatcher.dispatch(new KlighdFitToScreenAction(true));
+        }
+    }
+
+    handleLocally(action: Action): boolean {
+        // In contract to the name, this should return true, if the actions should be
+        // sent to the server. Don't know what the Sprotty folks where thinking when they named it...
+        switch (action.kind) {
+            case ComputedBoundsAction.KIND:
+                // TODO: remove sending of a computedBoundsAction as well
+                // (not possible until https://github.com/inversify/InversifyJS/issues/1035).
+                return false;
+            case ComputedTextBoundsAction.KIND:
+                return true;
+            case PerformActionAction.KIND:
+                return true;
+            case RequestTextBoundsCommand.KIND:
+                return false;
+            case SetSynthesisAction.KIND:
+                return true;
+            case RefreshLayoutAction.KIND:
+                return true;
+            case RefreshDiagramAction.KIND:
+                return true;
+        }
+        return super.handleLocally(action);
+    }
+
+    initialize(registry: ActionHandlerRegistry): void {
+        super.initialize(registry);
+
+        // Register the KLighD specific new actions.
+        registry.register(BringToFrontAction.KIND, this);
+        registry.register(CheckImagesAction.KIND, this);
+        registry.register(CheckedImagesAction.KIND, this);
+        registry.register(ComputedTextBoundsAction.KIND, this);
+        registry.register(DeleteLayerConstraintAction.KIND, this);
+        registry.register(DeletePositionConstraintAction.KIND, this);
+        registry.register(DeleteStaticConstraintAction.KIND, this);
+        registry.register(PerformActionAction.KIND, this);
+        registry.register(RectPackSetPositionConstraintAction.KIND, this);
+        registry.register(RectPackDeletePositionConstraintAction.KIND, this);
+        registry.register(RefreshDiagramAction.KIND, this);
+        registry.register(RefreshLayoutAction.KIND, this);
+        registry.register(RequestKlighdPopupModelAction.KIND, this);
+        registry.register(RequestTextBoundsCommand.KIND, this);
+        registry.register(SetAspectRatioAction.KIND, this);
+        registry.register(SetLayerConstraintAction.KIND, this);
+        registry.register(SetPositionConstraintAction.KIND, this);
+        registry.register(SetStaticConstraintAction.KIND, this);
+        registry.register(SetSynthesisAction.KIND, this);
+        registry.register(StoreImagesAction.KIND, this);
+        registry.register(SwitchEditModeAction.KIND, this);
+    }
+
+    handle(action: Action): void | ICommand | Action {
```

Comment:
As you might be aware, I introduced the concept of registries (state stores), which implement the `IActionHandler` interface. So where possible, actions sent from the server such as `SetSynthesesAction` moved from the `DiagramServer` handler to their respected registry.
The only action for simulation/compilation I am aware of, is `PerformActionAction` with the respected `actionId`s. The `DiagramServer` is completely unable to handle these cases and should not be able to, since it is out of its scope.
Previously, the `DiagramServer` "simply" notified event listeners that the action occurred. However, in a VS Code scenario, where the `DiagramServer` lives inside a webview (completely encapsulated and isolated) and the relevant simulation/compilation code is part of the extension, this architecture does not work since an extension can only listen for messages from the webview and is not able to interact with the webview in more direct ways.

Therefore, the `DiagramServer` is only really able to send the `PerformActionAction` towards the server (happens via the `Connection`). Spoiler: Which it does. Along the way, the action message bubbles through the extension, where `sprotty-vscode` provides a [mechanism to intercept actions](https://github.com/eclipse/sprotty-vscode/blob/master/sprotty-vscode-extension/src/sprotty-webview.ts#L203) and decide whether or not they should be send further along.
However, the `klighd-vscode` extension is also unaware of what should happen with the `PerformActionAction`, since simulation and compilation is not part of it's domain. Instead, it provides a mechanism for dependent extensions to register an action handler, which is implemented [here](https://github.com/kieler/klighd-vscode/blob/d0a0f6c6d6611fb0f94dc1ac11d189338de1555b/applications/klighd-vscode/src/extension.ts#L70-L92). So essentially the logic is delegated all the way to the `keith-vscode` extension which uses the provided mechanism to intercept simulation/compilation actions before they would finally be send to the server. The `keith-vscode` extension is finally the correct place where the domain should and can be implemented. The scaffold implementation can be found [here](https://git.rtsys.informatik.uni-kiel.de/projects/KIELER/repos/keith/browse/keith-vscode/src/perform-action-handler.ts?at=refs%2Fheads%2Fcfr%2Fmonorepo-restructure).

Reply:
When I wrote this comment I did not make it to those parts in the code yet. The handling looks good now, thanks for the clarification.